### DISCUSSION TITLE
plugin as output option

### DISCUSSION THREAD

**Discussion by itaysk** [aquasecurity/trivy#4451] [Ideas]
### Description

As most plugins are used to post-process trivy's output (format it/send it somewhere) we want allow invoking plugins as output options. This is supposed to address community requests to support more and more output formats and destinations, which can now be implemented by the community.

Before:

```
trivy image alpine | trivy plugin defectdojo
```

After:

```
trivy image alpine --output plugin=defectdojo
```

The mechanics remain the same in both use cases but the UX is more integrated.

1. A plugin declares itself as "output plugin", which means the plugin is relevant for the output flag (unlike for example kubectl trivy plugin which isn't). This done by setting  `type: output`.
2. A plugin declares which input format it expects, which means it will receive data in this format over stdin. This is done by setting `inputFormat:` to one of the available Trivy `--format` options.
3. (Note, need to bump plugin yaml's version)
4. `--output` flag now accepts a special `plugin=` notation which tells trivy to use a plugin. the value is the name of a preinstalled plugin. There can only be a single plugin specified.
5. output plugin can be configured by accepting command line flags. User can configure plugin flags by adding `key=value` comma separated values in the  `--output` value. For example `--output plugin=defectdojo,url=https://something` will invoke `trivy-defectdojo --url https://something`.
6. TBD: wait for process? error handling? can be discussed in PR

### Target

None

### Scanner

None

**Comment by fatihtokus** [comment]
This looks promising, any ETA?

**Comment by knqyf263** [comment]
I'd start with the minimum specification and grow it up gradually.

>1. A plugin declares itself as "output plugin", which means the plugin is relevant for the output flag (unlike for example kubectl trivy plugin which isn't). This done by setting type: output.

In the first version, any plugin can be called regardless of the type. It means we will not add `type: output`. We can add it later.

>2. A plugin declares which input format it expects, which means it will receive data in this format over stdin. This is done by setting inputFormat: to one of the available Trivy --format options.

To make it more flexible, I think `--format` should be specified in the command line rather than the plugin definition. We can make `--format json` the default with the output plugin. If the plugin wants to take SBOM as the input for example, they can configure it like `$ trivy image debian --format cyclonedx --output plugin=defectdojo`

>5. output plugin can be configured by accepting command line flags. User can configure plugin flags by adding key=value comma separated values in the --output value. For example --output plugin=defectdojo,url=https://something will invoke trivy-defectdojo --url https://something.

We may want to use the `args_`/`flags_` prefix or something like that. Then, we can make it a bit easier to support the comma-separated plugin args.

```
--output plugin=defectdojo,args_url=https://something,args_labels=key1=value1,key2=value2
```
